src/drt/src/ta/FlexTA_graphics.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2020, The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the University nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "FlexTA_graphics.h"
+
+#include "FlexTA.h"
+
+namespace fr {
+
+FlexTAGraphics::FlexTAGraphics(frDebugSettings* settings,
+                               frDesign* design,
+                               odb::dbDatabase* db,
+                               Logger* logger)
+    : logger_(logger),
+      settings_(settings),
+      gui_(gui::Gui::get()),
+      top_block_(design->getTopBlock())
+{
+  // Build the layer map between opendb & tr
+  auto odb_tech = db->getTech();
+
+  layer_map_.resize(odb_tech->getLayerCount(), -1);
+
+  for (auto& tr_layer : design->getTech()->getLayers()) {
+    auto odb_layer = odb_tech->findLayer(tr_layer->getName().c_str());
+    if (odb_layer) {
+      layer_map_[odb_layer->getNumber()] = tr_layer->getLayerNum();
+    }
+  }
+
+  gui_->registerRenderer(this);
+}
+
+void FlexTAGraphics::drawLayer(odb::dbTechLayer* layer, gui::Painter& painter)
+{
+  frLayerNum layerNum;
+
+  layerNum = layer_map_.at(layer->getNumber());
+  if (layerNum < 0) {
+    return;
+  }
+
+  if (settings_->netName.empty()) {
+    return;
+  }
+
+  for (auto& net : top_block_->getNets()) {
+    if (net->getName() != settings_->netName) {
+      continue;
+    }
```

Comment:
You could look up the net once rather than on every layer draw.  On a big design it might help.

Reply:
I'm wondering how best to cache the `FrNet *`. Would we have to handle updates, eg
```
set_detailed_route_debug -net net1
detailed_route
set_detailed_route_debug -net net2
detailed_route
```

I'm not sure if the detailed routing instance is reused or deleted and recreated.

Reply:
Just like you have a endIter callback you could have a start callback where you do the lookup.

Reply:
Makes sense. I just did it in endIter, since the data only gets written back to the database at the end of an iteration.

I also changed it to dump all iroutes if `-net` isn't specified, since I thought it might be useful to visualize.